text/3830-dedented-string-literals.md

```diff
@@ -0,0 +1,1296 @@
+    - Only the amount equal to the closing indentation, or less, will be removed.
+    - Never more than the line actually has.
+
+### Edge Cases
+
+> [!NOTE]
+>
+> `•` denotes a space.
+
+````rs
+// the whitespace at the start of non-empty lines is not part
+// of the calculation for "common indentation"
+// amongst non-empty lines
+//
+// remove the smallest amount of leading whitespace
+assert_eq!(
+    d"
+••••hello
+••
+••••world
+    ",
+^^^^ // common leading whitespace (will be removed)
+
+    "hello\nworld"
+);
+
+// line consisting of only spaces is allowed
+
+// However, nothing is removed because the:
+
+// > common indentation of all non-empty lines
+
+// is 0 here. (all lines are empty)
+
+// so min(0, x) = 0 -> remove 0 characters
+assert_eq!(
+    d"
+••••••••
+    ",
+
+    "••••••••"
+);
+
+// no whitespace removed either
+assert_eq!(
+    d"
+••••••••
+",
+
+    "••••••••"
+);
+
+// explanation:
+//
+// Initially we have:
+//
+// ```rust
+// let _ = d"
+//
+// ";
+// ```
+//
+// The literal newline directly after the opening `"` is removed. We get:
+//
+// ```rust
+// let _ = "
+// ";
+// ```
+//
+// The literal newline directly before the line containing
+// the closing `"` is removed. We get:
+//
+// ```rust
+// let _ = "";
+// ```
+//
+// An empty string.
+assert_eq!(
+    d"
+
+",
+
+    ""
+);
+
+// error: Expected a literal newline character
+//        before the line containing the closing quote
+//
+// note: The literal newline character after the opening quote
+//       is removed in all cases
+#[expect_compile_error]
+let _ = d"
+    ",
+````
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Contributes to the increase of string literal modifiers by adding a new variant.
+
+  While at the moment the variety of string literal modifiers is small, it is worth to think about the implications of exponential increase of them.
+
+  Currently, Rust has 7 types of string literals. This RFC will increase that to 13, because each string literal can be prefixed with a `d` to make it dedented.
+
+  In the future Rust might get additional types of "string modifiers", and each combination will need to
+  be accounted for.
+
+- Increases complexity of the language. While it builds upon existing concepts, it is yet another thing for people to learn.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Design
+
+### The choice of `d"string"` specifically
+
+The syntax of `d"string"` is chosen for the following reasons:
+
+- Fits with existing string modifiers, such as `b"string"`, `r#"string"#"` and `c"string"`
+- Composes with existing string modifiers: `db"string"`, `dc"string"`, `dr#"string"#`, and `dbr#"string"#`. 
+- Does not introduce a lot of new syntax. Dedented string literals can be explained in terms of existing language features.
+- The acronym `d` for `dedent` is both clear, and not taken by any of the other string modifiers.
```

Comment:
Counterexample: `u` for unindent.

I agree that it's understandable, but I wouldn't necessarily say that it's clear; it is definitely something that is learned and then understood, rather than immediately obvious by itself, unlike something like `c` which is immediately obvious as a C string, or `b` which is immediately obvious as a byte string. In both of those cases, you can know of the construct, see the literal syntax, and then connect the dots.

For dedented strings, there is no external Rust construct to connect with, and the word "dedent" itself, although common, is not powerful enough to make the meaning of `d"..."` obvious, IMHO.

Reply:
Other possibilities include `i"..."` for "indented string". That has the advantage of `i` at least being evocative of "indentation".

(Not looking to bikeshed, just to sketch some possible design requirements related to being memorable and intuitive.)

Reply:
FWIW, my intent wasn't to bikeshed, but rather point out that I don't think that "clarity" is necessarily one of the strong points of this design, regardless of what letter we choose. That doesn't mean it's a bad design, just that this isn't a slam dunk in this particular aspect.

Reply:
Drive-by mostly-bikeshed-sorry: were this accepted, I would specifically avoid `u"..."` because of previous use of `u` as a prefix meaning Unicode string in other languages (eg Python, especially Python 2). I suspect a careless reader in a polyglot environment may be prone to misinterpreting that specific prefix.

Reply:
On `u`: Python 2 has used the `u` prefix as a prefix for Unicode strings, and C++ currently uses it for UTF-16 strings, so I think `u` would lead to confusion here.

Reply:
(Also worth noting that I wasn't proposing `u`, since I agree that it's too confusing, even though Rust strings are (almost) always Unicode.)